client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,269 @@
+	/// Leaves should have already been ordered by "freshness" (older first).
+	fn remove_block(&mut self, number: NumberFor<Block>, leaves: &[Block::Hash]) {
+		let mut candidate_freshest_leaf_idx = usize::MAX;
+		let mut candidate_freshest_route = None;
+
+		let blockchain = self.backend.blockchain();
+		let best_hash = blockchain.info().best_hash;
+
+		for blk_hash in self.levels.entry(number).or_default().iter() {
+			// Start checking from the most fresh leaf (the last). Because of the leaves ordering,
+			// here leaf index is synonym of its freshness, that is the greater the index the
+			// fresher it is.
+			for (leaf_idx, leaf_hash) in leaves.iter().enumerate().rev() {
+				match sp_blockchain::tree_route(blockchain, *blk_hash, *leaf_hash) {
+					Ok(route) if route.retracted().is_empty() => {
+						// Skip this block if it is the ancestor of the best block or if it has
+						// a descendant leaf fresher (or equally fresh) than the current candidate.
+						if leaf_idx < candidate_freshest_leaf_idx &&
+							route.common_block().hash != best_hash &&
+							route.enacted().iter().all(|entry| entry.hash != best_hash)
+						{
+							// We have a new candidate
+							candidate_freshest_leaf_idx = leaf_idx;
+							candidate_freshest_route = Some(route);
+						}
+						break
+					},
+					Err(err) => {
+						log::warn!(
+							target: "parachain",
+							"Unable getting route from {} to {}: {}",
+							blk_hash, leaf_hash, err,
+						);
+					},
+					_ => (),
+				};
+			}
+			if candidate_freshest_leaf_idx == 0 {
+				// We can't find a candidate with an older leaf.
+				break
+			}
+		}
+
+		let candidate_freshest_route = match candidate_freshest_route {
+			Some(route) => route,
+			None => return,
+		};
+
+		// We have a candidate, proceed removing the block and all its descendants.
+
+		let candidate_hash = candidate_freshest_route.common_block().hash;
+
+		// Takes care of route removal. Starts from the leaf and stops as soon as an error is
+		// encountered. In this case an error is interpreted as the block being not a leaf
+		// and it will be removed while removing another route from the same block but to a
+		// different leaf.
+		let mut remove_route = |route: TreeRoute<Block>| {
+			std::iter::once(route.common_block()).chain(route.enacted()).rev().all(|elem| {
+				if self.backend.remove_leaf_block(&elem.hash).is_err() {
+					return false
+				}
+				log::debug!(target: "parachain", "Removing block {}", elem.hash);
+				self.levels
+					.get_mut(&elem.number)
+					.and_then(|level| level.remove(&elem.hash).then_some(()));
+				self.freshness.remove(&elem.hash);
+				true
+			});
+		};
+
+		remove_route(candidate_freshest_route);
+
+		let to_skip = leaves.len() - candidate_freshest_leaf_idx;
+		leaves.iter().rev().skip(to_skip).for_each(|leaf_hash| {
+			match sp_blockchain::tree_route(blockchain, candidate_hash, *leaf_hash) {
+				Ok(route) if route.retracted().is_empty() => {
+					remove_route(route);
```

Comment:
> None of this guarantees that you remove a block at level number.

Why? If there are not bugs somewhere in the backend this code should end up removing something 

---

The strategy in my implementation is:

> The freshness of each block at level `number` is equal to the freshness of its freshest leaf 

Sorry about all this freshness words :-D 

I've chosen this strategy because the sum may end up choosing not the best candidate (IMO).

For example lets assume I have this situation (attached the blocks freshness):

```
Level:   |   1   |   2   |   3   |   4   |
            
            A(1)----B(2)----C(3)----G(7)
              \
               +----D(4)----E(5)
                     \
                      +-----F(6)
```

If we want to remove something at level 2
- The "sum" strategy suggests to remove B (SUM @ D = 15, SUM for B = 12)
- The strategy I've implemented suggests to remove D  (fresher leaf @ D = 6, fresher leaf @ B = 7)


---

Note that I've updated the code and I've isolated the removal code from the lookup code.
In this way we can eventually change the target search code easily